File: Actividad 01. Sockets/Cliente en Python/cliente.py
import socket
import time

# Constantes

MIN_LONGITUD_USUARIO = 5
MAX_LONGITUD_USUARIO = 15
MIN_LONGITUD_CONTRASENA = 8
MAX_LONGITUD_CONTRASENA = 50

IP_SERVIDOR = "127.0.0.1"
PUERTO_SERVIDOR = 8888

BUFFER_SIZE = 1024

MAX_REINTENTOS = 5
REINTENTO_DELAY = 5  # en segundos

def mostrar_menu():
    """
    Muestra el menú al usuario y devuelve la opción ingresada como entero.

    :return: La opción ingresada por el usuario (1, 2 o 3).
    """

    print("Menú:")
    print("1. Generar nombre de usuario")
    print("2. Generar password")
    print("3. Salir")
    return int(input("Ingrese la opción deseada: "))

def enviar_datos(sock, mensaje):
    try:
        sock.sendall(mensaje.encode())
        return True
    except socket.error as e:
        print(f"Error al enviar datos: {e}")
        return False

def recibir_datos(sock):
    try:
        datos = sock.recv(BUFFER_SIZE).decode()
        if not datos:
            print("El servidor se desconectó.")
            return None
        return datos
    except socket.error as e:
        print(f"Error al recibir datos: {e}")
        return None

def intentar_reconexion(sock, direccion_servidor):
    
    """
    Intenta reconectar al servidor.

    :param socket sock: El socket a usar para la reconexión.
    :param tuple direccion_servidor: La dirección del servidor (IP, puerto).

    :return: True si se logra la reconexión, False en caso contrario.
    """

    intento = 0
    while intento < MAX_REINTENTOS:
        try:
            sock.connect(direccion_servidor)
            print("Reconexión exitosa.")
            return True
        except socket.error:
            print(f"Reintento de conexión en {REINTENTO_DELAY} segundos...")
            time.sleep(REINTENTO_DELAY)
            intento += 1
    return False

def procesar_opcion(opcion, sock, direccion_servidor):

    """
    Procesa la opción ingresada por el usuario.

    :param int opcion: La opción ingresada por el usuario (1, 2 o 3).
    :param socket sock: El socket para enviar los datos al servidor.
    :param tuple direccion_servidor: La dirección del servidor (IP, puerto).

    :return: None
    """
    
    if opcion == 1:
        longitud = int(input(f"Ingrese la longitud del nombre de usuario ({MIN_LONGITUD_USUARIO}-{MAX_LONGITUD_USUARIO}): "))
        mensaje = f"USER {longitud}"
    elif opcion == 2:
        longitud = int(input(f"Ingrese la longitud de la password ({MIN_LONGITUD_CONTRASENA}-{MAX_LONGITUD_CONTRASENA}): "))
        mensaje = f"PASS {longitud}"
    elif opcion == 3:
        print("Saliendo...")
        sock.close()
        exit(0)
    else:
        print("Opción no válida.")
        return

    if enviar_datos(sock, mensaje):
        respuesta = recibir_datos(sock)
        if respuesta:
            print(f"Respuesta del servidor: {respuesta}")
        else:
            print("Se perdió la conexión con el servidor. Intentando reconectar...")
            if intentar_reconexion(sock, direccion_servidor):
                print("Reconexión exitosa.")
            else:
                print("No se pudo reconectar al servidor. Saliendo...")
                sock.close()
                exit(1)

def main():

    """
    Punto de entrada principal del cliente.

    Establece una conexión con el servidor y luego entra en un bucle
    donde se muestra el menú al usuario y se procesa la opción ingresada.

    :return: None
    """

    direccion_servidor = (IP_SERVIDOR, PUERTO_SERVIDOR)
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    try:
        direccion_servidor = (IP_SERVIDOR, PUERTO_SERVIDOR)
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

        try:
            sock.connect(direccion_servidor)
            print("Conexión establecida.\n")
        except socket.error as e:
            print(f"Fallo la conexión con el servidor: {e}")
            sock.close()
            return

        while True:
            opcion = mostrar_menu()
            print("\n")
            procesar_opcion(opcion, sock, direccion_servidor)
            print("\n")

    except Exception as e:
        print(f"Ocurrió un error inesperado: {e}")
    finally:
        sock.close()
        print("\n")
        input("Presiona Enter para salir...")

if __name__ == "__main__":
    main()

File: Actividad 01. Sockets/Servidor y Cliente en C/cliente.c
#include <stdio.h>
#include <stdlib.h>
#include <windows.h>
#include <winsock2.h>
#include "cliente.h"
#include "socket_utils.h"
#include "config.h"

/**
 * Muestra el menú principal al usuario.
 *
 * Imprime un menú con las opciones disponibles y pide al usuario
 * que ingrese la opción deseada.
 */
void mostrarMenu(void)
{
    printf("Menu:\n");
    printf("1. Generar nombre de usuario\n");
    printf("2. Generar password\n");
    printf("3. Salir\n");
    printf("Ingrese la opcion deseada: ");
}

/**
 * Procesa la opción ingresada por el usuario y envía una solicitud al servidor.
 *
 * Dependiendo de la opción seleccionada, solicita al usuario la longitud del nombre de usuario
 * o la contraseña, y luego envía la solicitud correspondiente al servidor.
 *
 * @param opcion La opción ingresada por el usuario (1 para nombre de usuario, 2 para contraseña, 3 para salir).
 * @param socketCliente El socket del cliente utilizado para comunicarse con el servidor.
 * @param direccionServidor Estructura que contiene la dirección IP y el puerto del servidor.
 */
void procesarOpcion(int opcion, SOCKET socketCliente, struct sockaddr_in *direccionServidor)
{
    char buffer[BUFFER_SIZE];
    int longitud;

    switch (opcion)
    {
    case 1:
        printf("Ingrese la longitud del nombre de usuario (%d-%d): ", MIN_LONGITUD_USUARIO, MAX_LONGITUD_USUARIO);
        scanf("%d", &longitud);
        snprintf(buffer, sizeof(buffer), "USER %d", longitud);
        break;
    case 2:
        printf("Ingrese la longitud de la password (%d-%d): ", MIN_LONGITUD_CONTRASENA, MAX_LONGITUD_CONTRASENA);
        scanf("%d", &longitud);
        snprintf(buffer, sizeof(buffer), "PASS %d", longitud);
        break;
    case 3:
        printf("Saliendo...\n");
        exit(0);
    default:
        printf("Opcion no valida.\n");
        return;
    }

    if (enviarDatos(socketCliente, buffer) == EXIT_SUCCESS)
    {
        if (recibirDatos(socketCliente, buffer, sizeof(buffer)) == EXIT_SUCCESS)
        {
            printf("Respuesta del servidor: %s\n", buffer);
        }
    }
    else
    {
        printf("Se perdio la conexion con el servidor. Intentando reconectar...\n");
        if (intentarReconexion(&socketCliente, direccionServidor) == EXIT_SUCCESS)
        {
            printf("Reconexion exitosa.\n");
        }
        else
        {
            printf("No se pudo reconectar al servidor. Saliendo...\n");
            exit(EXIT_FAILURE);
        }
    }
}

/* CLIENTE */
int main()
{
    WSADATA datosWinsock;
    SOCKET socketCliente;
    struct sockaddr_in direccionServidor;
    int opcion;
    char mensaje[BUFFER_SIZE]; // Tamaño del búfer para el mensaje del usuario

    // Solicitar mensaje al usuario
    /*
    printf("Ingrese el mensaje a enviar: ");
    if (fgets(mensaje, sizeof(mensaje), stdin) == NULL)
    {
        fprintf(stderr, "Error al leer el mensaje.\n");
        return EXIT_FAILURE;
    }
    */

    // Eliminar el carácter de nueva línea si existe
    size_t len = strlen(mensaje);
    if (len > 0 && mensaje[len - 1] == '\n')
    {
        mensaje[len - 1] = '\0';
    }

    // Inicializar Winsock
    if (inicializarWinsock(&datosWinsock) != 0)
    {
        return EXIT_FAILURE;
    }

    // Crear el socket
    if ((socketCliente = crearSocket()) == INVALID_SOCKET)
    {
        manejarErrorYLimpiar(socketCliente, &datosWinsock);
    }

    // Configurar la dirección del socket
    configurarServidor(&direccionServidor, IP_SERVIDOR, PUERTO_SERVIDOR);

    // Conectar al servidor
    if (conectarAlServidor(socketCliente, &direccionServidor) != EXIT_SUCCESS)
    {
        manejarErrorYLimpiar(socketCliente, &datosWinsock);
    }

    // Enviar datos al servidor
    /*
    if (enviarDatos(socketCliente, mensaje) != EXIT_SUCCESS)
        manejarErrorYLimpiar(socketCliente, &datosWinsock);
    */

    while (1)
    {
        mostrarMenu();
        scanf("%d", &opcion);
        printf("\n");
        procesarOpcion(opcion, socketCliente, &direccionServidor);
        printf("\n");
    }

    // Limpiar y cerrar el socket
    closesocket(socketCliente);
    WSACleanup();

    system("pause");

    return EXIT_SUCCESS;
}
